vm/devices/pci/pci_core/src/capabilities/pci_express.rs

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! PCI Express Capability with Function Level Reset (FLR) support.
+
+use super::PciCapability;
+use crate::spec::caps::CapabilityId;
+use crate::spec::caps::pci_express;
+use crate::spec::caps::pci_express::PciExpressCapabilityHeader;
+use inspect::Inspect;
+use parking_lot::Mutex;
+use std::sync::Arc;
+
+/// FLR bit is the 28th bit in the Device Capabilities register (0 indexed).
+pub const PCI_EXPRESS_DEVICE_CAPS_FLR_BIT_MASK: u32 = 1 << 28;
+
+/// Callback interface for handling Function Level Reset (FLR) events.
+pub trait FlrHandler: Send + Sync + Inspect {
+    /// Called when Function Level Reset is initiated.
+    fn initiate_flr(&self);
+}
+
+#[derive(Debug, Inspect)]
+struct PciExpressState {
+    device_control: pci_express::DeviceControl,
+    device_status: pci_express::DeviceStatus,
+}
+
+impl PciExpressState {
+    fn new() -> Self {
+        Self {
+            device_control: pci_express::DeviceControl::new(),
+            device_status: pci_express::DeviceStatus::new(),
+        }
+    }
+}
+
+#[derive(Inspect)]
+/// Configurable PCI Express capability.
+pub struct PciExpressCapability {
+    device_capabilities: pci_express::DeviceCapabilities,
+    state: Arc<Mutex<PciExpressState>>,
+    #[inspect(skip)]
+    flr_handler: Option<Arc<dyn FlrHandler>>,
+}
+
+impl PciExpressCapability {
+    /// Creates a new PCI Express capability with FLR support.
+    ///
+    /// # Arguments
+    /// * `flr_handler` - Optional handler to be called when FLR is initiated. FLR support will be inferred if flr_handler = Some(_)
+    pub fn new(flr_handler: Option<Arc<dyn FlrHandler>>) -> Self {
+        Self {
+            device_capabilities: pci_express::DeviceCapabilities::new()
+                .with_function_level_reset(flr_handler.is_some()),
+            state: Arc::new(Mutex::new(PciExpressState::new())),
+            flr_handler,
+        }
+    }
+
+    fn handle_device_control_write(&mut self, new_control: pci_express::DeviceControl) {
+        let mut state = self.state.lock();
+
+        // Check if FLR was initiated
+        let old_flr = state.device_control.initiate_function_level_reset();
+        let new_flr = new_control.initiate_function_level_reset();
+
+
+        // DEVNOTE: It is "safe" to drop a new FLR request if there is still a previous FLR request in progress. The PCIe spec indicates that such behavior is undefined, so we choose to ignore the new FLR request.
```

Comment:
Please wrap at ~80 (we need a linter for this...)

Reply:
Do we wrap at 80? Most of our other code bases have moved to 120 or more.

Reply:
We conventionally wrap comments at 80 for readability. Code we leave up to rustfmt defaults.